src/MSBuild/XMake.cs

```diff
@@ -1255,27 +1255,66 @@ private static BuildResult ExecuteBuild(BuildManager buildManager, BuildRequestD
             ProjectCollection projectCollection,
             Dictionary<string, string> globalProperties)
         {
-            lock (s_buildLock)
-            {
-                var projectGraph = new ProjectGraph(projectFile, projectCollection, globalProperties);
-                var targetLists = projectGraph.GetTargetLists(targets);
+            // TODO: Handle Ctrl-C cancellation properly.
+            // s_activeBuild is supposed to be the active build, but we may have multiple going at once.
+            // This may be solved once better integration with the scheduler (#3774) is done.
+            var projectGraph = new ProjectGraph(projectFile, projectCollection, globalProperties);
+            var targetLists = projectGraph.GetTargetLists(targets);
 
-                // TODO: Do a full graph traversal
-                var project = projectGraph.EntryProjectNode.Project;
-                var targetList = targetLists[projectGraph.EntryProjectNode];
-                var request = new BuildRequestData(project, targetList.ToArray());
+            var waitHandle = new AutoResetEvent(true);
+            var graphBuildStateLock = new object();
 
-                s_activeBuild = buildManager.PendBuildRequest(request);
+            var blockedNodes = new HashSet<ProjectGraphNode>(projectGraph.ProjectNodes);
+            var finishedNodes = new HashSet<ProjectGraphNode>(projectGraph.ProjectNodes.Count);
+            var buildingNodes = new Dictionary<BuildSubmission, ProjectGraphNode>();
+            BuildResult entryProjectResult = null;
+            while (blockedNodes.Count > 0 || buildingNodes.Count > 0)
+            {
+                waitHandle.WaitOne();
 
-                // Even if Ctrl-C was already hit, we still pend the build request and then cancel.
-                // That's so the build does not appear to have completed successfully.
-                if (s_buildCancellationSource.IsCancellationRequested)
+                lock (graphBuildStateLock)
                 {
-                    buildManager.CancelAllSubmissions();
+                    var unblockedNodes = blockedNodes
+                        .Where(node => node.ProjectReferences.All(projectReference => finishedNodes.Contains(projectReference)))
```

Comment:
Yea, this is just an initial impl for now. My hope is that with #3774 we can just pend builds for all projects at once and express the dependencies via the existing "blocked by" mechanism in the scheduler (with minor tweaks since today I think the collection is "nodes blocked by this build request" rather than "build requests blocked by this build request") and just let it run. So the big O will be as good (or bad) as the existing scheduler :)